ring/replication_set_tracker.go

```diff
@@ -465,3 +466,76 @@ func (t *zoneAwareContextTracker) cancelAllContexts(cause error) {
 		delete(t.cancelFuncs, instance)
 	}
 }
+
+type inflightInstanceTracker struct {
+	mx       sync.Mutex
+	inflight []map[*InstanceDesc]struct{}
```

Comment:
nit: assuming that each RS will have 2-3 instances, maybe a slice would have been more efficient

Reply:
Will give it a try.

Reply:
Yes! It's better: https://github.com/grafana/dskit/pull/498